File: ios/DesMoinesInsider/Services/DeepLinkHandler.swift
import Foundation
import os

/// Parses deep links and universal links into app navigation destinations.
///
/// Supported URL patterns:
/// - `desmoinesinsider.com/events/:id` → Event detail
/// - `desmoinesinsider.com/restaurants/:id` → Restaurant detail
/// - `desmoinesinsider.com/attractions/:id` → Attraction detail
/// - `com.desmoines.aipulse://event/:id` → Event detail (custom scheme)
/// - `com.desmoines.aipulse://restaurant/:id` → Restaurant detail (custom scheme)
/// - `com.desmoines.aipulse://auth-callback` → Auth callback (handled by Supabase)
@MainActor
@Observable
final class DeepLinkHandler {
    static let shared = DeepLinkHandler()

    private(set) var pendingDestination: Destination?

    enum Destination: Equatable {
        case event(id: String)
        case restaurant(id: String)
        case attraction(id: String)
        case tab(MainTabView.Tab)
        /// A Discover-hub parity surface (IOS-IA-002), e.g. trip planner, deals.
        case discover(DiscoverDestination)
    }

    private init() {}

    // MARK: - Parse URL

    /// Attempts to parse a URL into a navigation destination.
    /// Returns `true` if the URL was handled, `false` if it should be passed to Supabase.
    @discardableResult
    func handle(_ url: URL) -> Bool {
        // Skip auth callbacks — let Supabase handle those
        if url.absoluteString.contains("auth-callback") {
            return false
        }

        if let destination = parseUniversalLink(url) ?? parseCustomScheme(url) {
            pendingDestination = destination
            return true
        }

        return false
    }

    func consumeDestination() -> Destination? {
        defer { pendingDestination = nil }
        return pendingDestination
    }

    // MARK: - Notifications (IOS-AUDIT-FEAT-003)

    /// Routes a notification payload to a destination. Event-reminder local
    /// notifications carry `eventId`; push payloads may carry a deep-link `url`
    /// or a typed `type`/`id` pair. Returns true if a destination was set.
    @discardableResult
    func handleNotification(userInfo: [AnyHashable: Any]) -> Bool {
        if let eventId = userInfo["eventId"] as? String,
           let id = validatedId(eventId, source: "notification") {
            pendingDestination = .event(id: id)
            return true
        }
        if let urlString = userInfo["url"] as? String, let url = URL(string: urlString) {
            return handle(url)
        }
        if let type = userInfo["type"] as? String, let rawId = userInfo["id"] as? String {
            return routeTyped(type: type, rawId: rawId)
        }
        return false
    }

    // MARK: - Spotlight (IOS-AUDIT-FEAT-027)

    /// Routes a Spotlight result tap. CoreSpotlight delivers the indexed item's
    /// `uniqueIdentifier` (e.g. "event-<uuid>") via CSSearchableItemActivityIdentifier.
    /// Parses the "<type>-<id>" form SpotlightService writes and routes the three
    /// content types that have detail destinations. Returns true if handled.
    @discardableResult
    func handleSpotlightIdentifier(_ identifier: String) -> Bool {
        guard let dash = identifier.firstIndex(of: "-") else { return false }
        let type = String(identifier[..<dash])
        let rawId = String(identifier[identifier.index(after: dash)...])
        switch type {
        case "event", "restaurant", "attraction":
            return routeTyped(type: type, rawId: rawId)
        default:
            // article/hotel are indexed but have no detail destination yet —
            // open the app without crashing rather than route to the wrong place.
            AppLogger.nav.warning("Unrouted Spotlight identifier type: \(type)")
            return false
        }
    }

    private func routeTyped(type: String, rawId: String) -> Bool {
        switch type {
        case "event":
            guard let id = validatedId(rawId, source: "notification") else { return false }
            pendingDestination = .event(id: id)
        case "restaurant":
            guard let id = validatedId(rawId, source: "notification") else { return false }
            pendingDestination = .restaurant(id: id)
        case "attraction":
            guard let id = validatedId(rawId, source: "notification") else { return false }
            pendingDestination = .attraction(id: id)
        default:
            return false
        }
        return true
    }

    // MARK: - ID Validation

    /// Validates that an ID looks like a UUID (8-4-4-4-12 hex format).
    /// Rejects malformed IDs that could cause unexpected behavior.
    private func isValidId(_ id: String) -> Bool {
        // UUID format: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
        UUID(uuidString: id) != nil
    }

    /// Validates and returns the ID, or nil if invalid (logging the rejection).
    private func validatedId(_ id: String, source: String) -> String? {
        if isValidId(id) { return id }
        AppLogger.nav.warning("Rejected invalid deep link ID from \(source): \(id.prefix(50))")
        return nil
    }

    // MARK: - Parse Helpers

    private func parseUniversalLink(_ url: URL) -> Destination? {
        guard let host = url.host,
              host.contains("desmoinesinsider.com") else { return nil }

        let path = url.pathComponents.filter { $0 != "/" }

        guard path.count >= 2 else {
            // Root path — navigate to appropriate tab
            if path.first == "events" { return .tab(.home) }
            if path.first == "restaurants" { return .tab(.restaurants) }
            // Single-segment Discover parity surfaces, e.g. /trip-planner.
            if let slug = path.first, let discover = DiscoverDestination(slug: slug) {
                return .discover(discover)
            }
            return nil
        }

        let type = path[0]
        let rawId = path[1]

        switch type {
        case "events":
            guard let id = validatedId(rawId, source: "universal-link") else { return .tab(.home) }
            return .event(id: id)
        case "restaurants":
            guard let id = validatedId(rawId, source: "universal-link") else { return .tab(.restaurants) }
            return .restaurant(id: id)
        case "attractions":
            guard let id = validatedId(rawId, source: "universal-link") else { return .tab(.home) }
            return .attraction(id: id)
        default:
            // Discover-hub parity surfaces (IOS-IA-002): the path's first
            // component is itself the slug, e.g. /trip-planner, /deals.
            if let discover = DiscoverDestination(slug: type) {
                return .discover(discover)
            }
            return nil
        }
    }

    private func parseCustomScheme(_ url: URL) -> Destination? {
        guard url.scheme == Config.appBundleId else { return nil }

        let host = url.host ?? ""
        let path = url.pathComponents.filter { $0 != "/" }
        let rawId = path.first ?? ""

        switch host {
        case "event" where !rawId.isEmpty:
            guard let id = validatedId(rawId, source: "custom-scheme") else { return .tab(.home) }
            return .event(id: id)
        case "restaurant" where !rawId.isEmpty:
            guard let id = validatedId(rawId, source: "custom-scheme") else { return .tab(.restaurants) }
            return .restaurant(id: id)
        case "attraction" where !rawId.isEmpty:
            guard let id = validatedId(rawId, source: "custom-scheme") else { return .tab(.home) }
            return .attraction(id: id)
        case "home": return .tab(.home)
        case "search": return .tab(.search)
        case "favorites": return .tab(.favorites)
        case "profile": return .tab(.profile)
        case "discover":
            // com.desmoines.aipulse://discover/<slug> (IOS-IA-002). Bare
            // //discover opens the hub's first tile's surface is not assumed;
            // require an explicit slug.
            if let discover = DiscoverDestination(slug: rawId) { return .discover(discover) }
            return nil
        default: return nil
        }
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/util/DeepLinkHandler.kt
package com.desmoines.aipulse.util

import android.content.Intent
import android.net.Uri
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Parses deep links and universal links into app navigation destinations.
 * Mirrors iOS DeepLinkHandler.swift.
 *
 * Supported URL patterns:
 * - `desmoinesinsider.com/events/:id` -> Event detail
 * - `desmoinesinsider.com/restaurants/:id` -> Restaurant detail
 * - `desmoinesinsider.com/attractions/:id` -> Attraction detail
 * - `com.desmoines.aipulse://event/:id` -> Event detail (custom scheme)
 * - `com.desmoines.aipulse://restaurant/:id` -> Restaurant detail (custom scheme)
 * - `com.desmoines.aipulse://attraction/:id` -> Attraction detail (custom scheme)
 * - `com.desmoines.aipulse://home` -> Home tab
 * - `com.desmoines.aipulse://search` -> Search tab
 * - `com.desmoines.aipulse://favorites` -> Saved tab
 * - `com.desmoines.aipulse://profile` -> Profile tab
 * - `com.desmoines.aipulse://auth-callback` -> Auth callback (handled by Supabase)
 *
 * Also handles notification tap intents with navigate_to + event_id/content_id extras.
 */
@Singleton
class DeepLinkHandler @Inject constructor() {

    sealed class Destination {
        data class Event(val id: String) : Destination()
        data class Restaurant(val id: String) : Destination()
        data class Attraction(val id: String) : Destination()
        data class Tab(val tab: TabDestination) : Destination()
    }

    enum class TabDestination {
        HOME, DINING, SEARCH, MAP, SAVED, PROFILE
    }

    private val _pendingDestination = MutableStateFlow<Destination?>(null)
    val pendingDestination: StateFlow<Destination?> = _pendingDestination.asStateFlow()

    /**
     * Attempts to parse a URL into a navigation destination.
     * Returns true if the URL was handled, false if it should be passed to Supabase (e.g., auth callbacks).
     */
    fun handle(url: Uri): Boolean {
        // Skip auth callbacks - let Supabase handle those
        if (url.toString().contains("auth-callback")) {
            return false
        }

        val destination = parseUniversalLink(url) ?: parseCustomScheme(url)
        if (destination != null) {
            _pendingDestination.value = destination
            return true
        }

        return false
    }

    /**
     * Handles intents from notification taps and other internal deep links.
     * Checks for navigate_to + event_id/content_id extras from NotificationReceiver/PushNotificationService.
     */
    fun handleIntent(intent: Intent): Boolean {
        // Check for notification tap extras
        val navigateTo = intent.getStringExtra("navigate_to")
        val contentId = intent.getStringExtra("event_id")
            ?: intent.getStringExtra("content_id")

        if (navigateTo != null) {
            val destination = when (navigateTo) {
                "event_detail" -> contentId?.let { Destination.Event(it) }
                "restaurant_detail" -> contentId?.let { Destination.Restaurant(it) }
                "attraction_detail" -> contentId?.let { Destination.Attraction(it) }
                "home" -> Destination.Tab(TabDestination.HOME)
                "search" -> Destination.Tab(TabDestination.SEARCH)
                "favorites" -> Destination.Tab(TabDestination.SAVED)
                "profile" -> Destination.Tab(TabDestination.PROFILE)
                else -> null
            }
            if (destination != null) {
                _pendingDestination.value = destination
                // Clear intent extras to prevent re-handling on config change
                intent.removeExtra("navigate_to")
                intent.removeExtra("event_id")
                intent.removeExtra("content_id")
                return true
            }
        }

        // Check for deep link URI in intent data
        val uri = intent.data
        if (uri != null) {
            return handle(uri)
        }

        return false
    }

    /**
     * Consumes and returns the pending destination, clearing it.
     */
    fun consumeDestination(): Destination? {
        val destination = _pendingDestination.value
        _pendingDestination.value = null
        return destination
    }

    // --- Parse helpers ---

    private fun parseUniversalLink(url: Uri): Destination? {
        val host = url.host ?: return null
        if (!host.contains("desmoinesinsider.com")) return null

        val pathSegments = url.pathSegments
        if (pathSegments.isNullOrEmpty()) return null

        // Single-segment paths: /events -> Home tab, /restaurants -> Dining tab
        if (pathSegments.size == 1) {
            return when (pathSegments[0]) {
                "events" -> Destination.Tab(TabDestination.HOME)
                "restaurants" -> Destination.Tab(TabDestination.DINING)
                else -> null
            }
        }

        // Two-segment paths: /events/:id, /restaurants/:id, /attractions/:id
        if (pathSegments.size >= 2) {
            val type = pathSegments[0]
            val id = pathSegments[1]
            if (id.isBlank()) return null

            return when (type) {
                "events" -> Destination.Event(id)
                "restaurants" -> Destination.Restaurant(id)
                "attractions" -> Destination.Attraction(id)
                else -> null
            }
        }

        return null
    }

    private fun parseCustomScheme(url: Uri): Destination? {
        val scheme = url.scheme ?: return null
        if (scheme != Config.APP_BUNDLE_ID) return null

        val host = url.host ?: return null
        val pathSegments = url.pathSegments
        val id = pathSegments?.firstOrNull() ?: ""

        return when {
            host == "event" && id.isNotEmpty() -> Destination.Event(id)
            host == "restaurant" && id.isNotEmpty() -> Destination.Restaurant(id)
            host == "attraction" && id.isNotEmpty() -> Destination.Attraction(id)
            host == "home" -> Destination.Tab(TabDestination.HOME)
            host == "search" -> Destination.Tab(TabDestination.SEARCH)
            host == "favorites" -> Destination.Tab(TabDestination.SAVED)
            host == "profile" -> Destination.Tab(TabDestination.PROFILE)
            host == "dining" -> Destination.Tab(TabDestination.DINING)
            host == "map" -> Destination.Tab(TabDestination.MAP)
            else -> null
        }
    }
}
